File: Objective-C/Multithreading in ObjectiveC/Multithreading in ObjectiveC/ViewController.m
//
//  ViewController.m
//  Multithreading in ObjectiveC
//
//  Created by Nigel Tan Yong on 2/3/24.
//

#import "ViewController.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UIImageView *imageView;
@end
@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}
- (IBAction)randomImage:(id)sender {
    NSURL *url = [NSURL URLWithString:@"https://loremflickr.com/2000/2000"];
    	
    dispatch_queue_t backgroundQueue = dispatch_get_global_queue(QOS_CLASS_BACKGROUND, 0);
    
    dispatch_async(backgroundQueue, ^{
        NSData *data = [NSData dataWithContentsOfURL:url];
        
        UIImage *image = [UIImage imageWithData: data];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            self.imageView.image = image;
        });
    });
}


@end

File: Swift/Mutithreading in Swift/Mutithreading in Swift/ViewController.swift
//
//  ViewController.swift
//  Mutithreading in Swift
//
//  Created by Nigel Tan Yong on 2/3/24.
//

import UIKit

class ViewController: UIViewController {
    @IBOutlet weak var imageView: UIImageView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }

    @IBAction func randomImage(_ sender: Any) {
        // Perform task in a background thread to avoid blocking the main UI thread
        DispatchQueue.global(qos: .background).async{
            // Construct URL for a random image
            let url = URL(string: "https://loremflickr.com/2000/2000")
            // Fetch image data from the URL
            let data = try! Data(contentsOf: url!)
            // Convert fetched data into UIImage
            let image = UIImage(data: data)
            
            // Update UI on the main thread
            DispatchQueue.main.async{
                self.imageView.image=image
            }
        }
    }
    
}

